### PR TITLE
Revamp Asset Create/Update Workflow using ShadCN, React-Hook-Form, and TanStack Query

### LINKED ISSUE #9189: Unnecessary API Calls in Assets Update Page
**Describe the bug**
In Assets update page, the form is allowing users to raise a PUT API call even when there's no change in the data, the same thing is present in in Service History update too

**To Reproduce**

1. Go to '/Assets'
2. Click on any asset
3. Select Update
4. Scroll Down and choose update without making any change
5. See Error

**Expected behavior**

Update Button should be deactived when there's no change in the data

**Screenshots**

https://github.com/user-attachments/assets/317eb18d-a5be-4a38-a74d-49e9c9b00e30

**Desktop**

- OS: Windows
- Browser Chrome
- Version Latest

### PR DESCRIPTION
**Pull request by rajku-dev**
## Proposed Changes

https://github.com/user-attachments/assets/2540789d-2152-4104-a877-20ca4fe3ddb3

- Fixes #9189 
-Initialized form state using 'fetchedValues' to pre-fill the form with asset data.
-Introduced a mechanism to compare current form values with initial values and enable/disable the submit button based on changes made by the user in form.
-Used a useRef (isInitialLoad.current) to skip comparisons during the initial load and avoid unnecessary state updates.
-Once the initial values are set, this flag is updated to ensure further state checks and updates are triggered properly to enable or disable the 'Update' button accordingly.

@ohcnetwork/care-fe-code-reviewers

## Merge Checklist

- [ ] Add specs that demonstrate bug / test a new feature.
- [ ] Update [product documentation](https://docs.ohc.network).
- [ ] Ensure that UI text is kept in I18n files.
- [ ] Prep screenshot or demo video for changelog entry, and attach it to issue.
- [ ] Request for Peer Reviews
- [ ] Completion of QA

## Release Notes

- **New Features**
  - Enhanced asset creation form with improved validation and form management.
  - Added new UI components for calendar, select, switch, textarea, and phone input.
  - Integrated advanced form handling with react-hook-form.

- **Dependencies**
  - Added new libraries for form management, UI components, and utilities.
  - Updated Zod validation library.

- **Improvements**
  - Streamlined form submission and error handling.
  - Improved user interface with more robust and flexible components.
  - Enhanced form validation with detailed schema checks.

### CODE REVIEW

File: src/components/Facility/AssetCreate.tsx
Comment (Jacobjeevan) [inline src/components/Facility/AssetCreate.tsx]: Let's switch to using Form Component.

**Review by Jacobjeevan — CHANGES_REQUESTED** [review]

File: src/components/Facility/AssetCreate.tsx
Comment (Jacobjeevan) [inline src/components/Facility/AssetCreate.tsx]: We don't want to display undefined in case name is not available; replace it with "Asset"

File: src/components/Facility/AssetCreate.tsx
Comment (Jacobjeevan) [inline src/components/Facility/AssetCreate.tsx]: Make sure to use names that make sense; AssetT doesn't tell us what it is.

Let's change it to AssetData instead.

File: src/components/Facility/AssetCreate.tsx
Comment (Jacobjeevan) [inline src/components/Facility/AssetCreate.tsx]: Switch rest of these to using translations as well, like in the line right above.

Add key value pair to en.json file. Some might already be present, so be sure to check first.

File: src/components/Facility/AssetCreate.tsx
Comment (Jacobjeevan) [inline src/components/Facility/AssetCreate.tsx]: #8956 adds an additionalButtons prop to Form Component, which will allow to keep the add more functionality. Take a look and write code (commented out) to work with it.

We can merge this PR once 8956 is merged.

File: src/components/Facility/AssetCreate.tsx
Comment (Jacobjeevan) [inline src/components/Facility/AssetCreate.tsx]: Update the state here (using setinitialAssetData....)

**Review by Jacobjeevan — CHANGES_REQUESTED** [review]

File: src/components/Facility/AssetCreate.tsx
Comment (rajku-dev) [reply src/components/Facility/AssetCreate.tsx]: @Jacobjeevan i suggest that `additionalButtons` prop  in #8956 should take `buttonDisable` property also, for each  button, in that way Create & Add More button will be enabled depending on if we have `assetId` or not,
although i have made necesssary changes in `AssetCreate`.

Comment (Jacobjeevan) [reply src/components/Facility/AssetCreate.tsx]: > @Jacobjeevan i suggest that `additionalButtons` prop in #8956 should take `buttonDisable` property also, for each button, in that way Create & Add More button will be enabled depending on if we have `assetId` or not, although i have made necesssary changes in `AssetCreate`.

Hmm, you can just make the value for `additionalButtons` conditional (go ahead and make the change because right now we don't display the save and add more when updating an existing asset).

Compare your changes to the form in care.ohc.network (you can use the dev-districtadmin login).

Note: I know form wouldn't work right now due to being blocked by 8956, but check as much as you can.

File: src/components/Facility/AssetCreate.tsx
Comment (Jacobjeevan) [reply src/components/Facility/AssetCreate.tsx]: Make this change as well.

Comment (rajku-dev) [reply src/components/Facility/AssetCreate.tsx]: > Make this change as well.

@Jacobjeevan done.

File: src/components/Facility/AssetCreate.tsx
Comment (rajku-dev) [reply src/components/Facility/AssetCreate.tsx]: @Jacobjeevan 
i need to pass buttonDisabled as of now? is that what you're saying?
may be like this
additionalButtons={[
                  {
                    type: "submit",
                    label: t("create_add_more"),
                    id: "create-asset-add-more-button",
                    _**buttonDisabled : assetId ? true : false,**_
                  },
   and existing submit handler 
onSubmit={async (obj) => {
                  await handleSubmitAsync(obj, _**Boolean(assetId)**_);
                }}
takes care of rest

Comment (Jacobjeevan) [reply src/components/Facility/AssetCreate.tsx]: > @Jacobjeevan i need to pass buttonDisabled as of now? is that what you're saying? may be like this additionalButtons={[ { type: "submit", label: t("create_add_more"), id: "create-asset-add-more-button", _**buttonDisabled : assetId ? true : false,**_ }, and existing submit handler onSubmit={async (obj) => { await handleSubmitAsync(obj, _**Boolean(assetId)**_); }} takes care of rest

No I meant, there is no need for additionalButtons to have anything if assetId is present (i.e. you are editing an existing asset).

Comment (rajku-dev) [reply src/components/Facility/AssetCreate.tsx]: > > @Jacobjeevan i need to pass buttonDisabled as of now? is that what you're saying? may be like this additionalButtons={[ { type: "submit", label: t("create_add_more"), id: "create-asset-add-more-button", _**buttonDisabled : assetId ? true : false,**_ }, and existing submit handler onSubmit={async (obj) => { await handleSubmitAsync(obj, _**Boolean(assetId)**_); }} takes care of rest
> 
> No I meant, there is no need for additionalButtons to have anything if assetId is present (i.e. you are editing an existing asset).


@Jacobjeevan Ah! i see
this right 
image

 but still, the addMore will be true if we don't have assetId, so both create and Create add more will have same logic (i.e they'll redirect to same page after creating the new asset)

currently i am able to see create asset is working same as create add more, once we have 8956 merged create add more will work same.

since the logic for submit is same for both buttons in create page, and only navigation after submission is dependent on addMore property in the arg of
onSubmit={async (obj) => {
                  await handleSubmitAsync(obj, !Boolean(assetId));
                }}
 currently what we are passing is not a right determinant of addMore!

Comment (rajku-dev) [reply src/components/Facility/AssetCreate.tsx]: one possible solution is this!

onSubmit={async (obj) => {
                  const clickedButtonId = (document.activeElement as HTMLButtonElement)?.id;
                  const addMore = clickedButtonId === "create-asset-add-more-button";
                  await handleSubmitAsync(obj, addMore);
                }}
or we can also access the label> using .textContent
this is working for Create Asset, lets see if it works when 8596 merges.

Comment (Jacobjeevan) [reply src/components/Facility/AssetCreate.tsx]: @rajku-dev That will work, but you don't need to handle the buttonId lookup either.

BedCapacity was doing something similar, so Form component had to be modified to support it.
https://github.com/ohcnetwork/care_fe/blob/20174b7531bacbc474e666bed6e7b21dd25277d2/src/components/Facility/BedCapacity.tsx#L159

https://github.com/ohcnetwork/care_fe/blob/20174b7531bacbc474e666bed6e7b21dd25277d2/src/components/Form/Form.tsx#L67 passes buttonId to the submit function, so you work based off that; i.e. change addMore param to be a check for specific the buttonId.

Comment (rajku-dev) [reply src/components/Facility/AssetCreate.tsx]: @Jacobjeevan got it!
                onSubmit={async (obj, buttonId) => {
                  await handleSubmitAsync(obj, buttonId == "create-asset-add-more-button");
                }}
               this right..

Comment (Jacobjeevan) [reply src/components/Facility/AssetCreate.tsx]: > @Jacobjeevan got it! onSubmit={async (obj, buttonId) => { await handleSubmitAsync(obj, buttonId == "create-asset-add-more-button"); }} this right..

You don't need to do that; form component would pass in the buttonId, just modify the addMore check accordingly inside the handleSubmit.

**Comment by Jacobjeevan** [comment]
Depends on #8956. Merge only after that PR is merged.

File: src/components/Facility/AssetCreate.tsx
Comment (rajku-dev) [reply src/components/Facility/AssetCreate.tsx]: @Jacobjeevan I think it is already handling 
try {
      if (!form.id) {
        const { res } = await request(routes.createAsset, { body: data });
        if (res?.ok) {
          Notification.Success({ msg: "Asset created successfully" });
          // Handle "Add More" logic 
          if (addMore) {
            resetFilters();
            const pageContainer = window.document.getElementById("pages");
            pageContainer?.scroll(0, 0);
          } else {
            goBack();
          }
        }
      } 
    Form comp will pass the id of button that is clicked, if it is the same as Create & Add More btn id i.e create-asset-add-more-button , addMore will be true, which is handled accordingly inside the handler.

Comment (Jacobjeevan) [reply src/components/Facility/AssetCreate.tsx]: @rajku-dev Yes, but to clarify, I meant moving the buttonId check inside handleSubmit, rather than passing it in.

Comment (rajku-dev) [reply src/components/Facility/AssetCreate.tsx]: > @rajku-dev Yes, but to clarify, I meant moving the buttonId check inside handleSubmit, rather than passing it in.

@Jacobjeevan done.

**Comment by rajku-dev** [comment]
> Depends on #8956. Merge only after that PR is merged.

@Jacobjeevan sir, can you please review the final changes?

File: src/components/Facility/AssetCreate.tsx
Comment (Jacobjeevan) [inline src/components/Facility/AssetCreate.tsx]: Add translations (here and elsewhere in this file). @rajku-dev 

Check en.json if there's already a value, if not add 👍

**Comment by Jacobjeevan** [comment]
> > Depends on #8956. Merge only after that PR is merged.
> 
> @Jacobjeevan sir, can you please review the final changes?

Nothing else apart from translations for now 👍 

Also, don't need to refer to anyone here as sir/madam. Just refer to us by our names.

File: src/components/Facility/AssetCreate.tsx
Comment (rithviknishad) [inline src/components/Facility/AssetCreate.tsx]: Let's not use relative imports. Switch to absolute imports.

File: src/components/Facility/AssetCreate.tsx
Comment (rithviknishad) [inline src/components/Facility/AssetCreate.tsx]: This type is already defined in our code base. Let's not redefine it. Import the existing type instead.

File: src/components/Facility/AssetCreate.tsx
Comment (rithviknishad) [inline src/components/Facility/AssetCreate.tsx]: Use the reusable `PhoneNumberValidator` instead.

File: src/components/Facility/AssetCreate.tsx
Comment (rithviknishad) [inline src/components/Facility/AssetCreate.tsx]:
```suggestion
  const handleSubmit = async (form: AssetData, buttonId: string) => {
```

File: src/components/Facility/AssetCreate.tsx
Comment (rithviknishad) [inline src/components/Facility/AssetCreate.tsx]: Switch to using `asyncGetDefaults` instead. This allows asynchronously fetching initial data and simplifies lot of the state management.

**Review by rithviknishad — CHANGES_REQUESTED** [review]
Optional but recommended: Since this is doing quite a lot of refactor on pretty much the entire asset form, I'd recommend switching to using shadcn's form components along with other shadcn's input components too.

File: src/components/Facility/AssetCreate.tsx
Comment (rajku-dev) [reply src/components/Facility/AssetCreate.tsx:88]: @rithviknishad existing AssetData type have serviced_on date and note encapsulated in last_service object, making it incompatible to be used with Form provider

File: src/components/Facility/AssetCreate.tsx
Comment (rajku-dev) [reply src/components/Facility/AssetCreate.tsx]: current code uses 
const assetQuery = useQuery(routes.getAsset, {
    pathParams: { external_id: assetId! },
    prefetch: !!assetId,
which is not async?

Comment (rithviknishad) [reply src/components/Facility/AssetCreate.tsx]: you can disable prefetch and call refetch in asyncGetDefaults

File: src/components/Facility/AssetCreate.tsx
Comment (rajku-dev) [reply src/components/Facility/AssetCreate.tsx]: @rithviknishad that type contains last_service object which has notes and on_service date encapsulated , but we have individual fields for those in form that uses Provider

**Comment by nihal467** [comment]
@rajku-dev can you share the status on this PR in the next 48 status, otherwise will close the PR and unassign you from the issue due to inactivity

**Comment by nihal467** [comment]
closing this PR due to inactivity

**Comment by rajku-dev** [comment]
Hi @nihal467,
Apologies for the inactivity on this PR. I’d love to pick it up again and continue working on it. Could you kindly reopen it if possible?

Additionally, I’m interested in exploring the new Care Do branch. Could you please guide me on where I can access it?

Looking forward to your response. Thank you! 😊

**Comment by Jacobjeevan** [comment]
> Hi @nihal467, Apologies for the inactivity on this PR. I’d love to pick it up again and continue working on it. Could you kindly reopen it if possible?
> 
> Additionally, I’m interested in contributing to the new Care Do branch. Could you please share the link or guide me on where I can explore its code?
> 
> Looking forward to your response. Thank you! 😊

image

You should be able to open the PR; check the comment box below 👍 

As for the new branch, check the `form-field-v1` (probably will be renamed to Do). You can use `REACT_CARE_API_URL=https://care-api.do.ohc.network/` in .env.local for calling the corresponding backend. https://github.com/ohcnetwork/care_fe/issues/9471 is the relevant bugs thread.

We aren't accepting contributions yet, but feel free to explore the code.

**Comment by rajku-dev** [comment]
Hi @jacobjeevan,

I checked but didn’t find that button. 
Thanks for the detailed information about the new branch. I’ll explore it

**Comment by Jacobjeevan** [comment]
> Hi @Jacobjeevan,
> 
> I checked but couldn’t find the `Reopen pull request` button. Could you clarify if I’m missing something?
> 
> Thanks for the detailed information about the new branch and backend setup. I’ll explore the `form-field-v1` branch and follow the thread for more insights.

Weird, I've reopened the PR for now.

**Comment by rajku-dev** [comment]
> Optional but recommended: Since this is doing quite a lot of refactor on pretty much the entire asset form, I'd recommend switching to using shadcn's form components along with other shadcn's input components too.

The recommendation is pretty good though!
Should we start working on it? @rithviknishad

**Comment by rithviknishad** [comment]
Yes, go ahead. Use shadcn's form. You can use it's `isDirty` check from useForm to disable the button instead.

**Comment by rajku-dev** [comment]
> Yes, go ahead. Use shadcn's form. You can use it's `isDirty` check from useForm to disable the button instead.


    "@radix-ui/react-icons": "^1.3.1",
    "@radix-ui/react-slot": "^1.1.0",
    "@radix-ui/react-toast": "^1.2.2",
    "@radix-ui/react-tooltip": "^1.1.3",
    
 there's no shadcn form here , are we supposed to use `react-hook-form` with existing `radix-ui` components
 @rithviknishad

**Comment by rithviknishad** [comment]
follow the docs: https://ui.shadcn.com/docs/components/form

**Comment by rajku-dev** [comment]
> follow the docs: https://ui.shadcn.com/docs/components/form

Should we replace all input components, such as DatePicker, Switch, Select, etc., with ShadCN components, as per the documentation?
https://ui.shadcn.com/docs/components @rithviknishad

**Comment by rithviknishad** [comment]
Closing as the related issue is closed as not planned.

**Comment by rajku-dev** [comment]
> Closing as the related issue is closed as not planned.

Hi @rithviknishad sir,
I’ve almost completed the AssetCreate form using Shadcn components, including the dirty state. The only task left is applying the tailwind classes, which I plan to finish today. I was about to make a commit tonight.
The form now restores all the original functionality, with improved state management and reduced clutter.

**Comment by rajku-dev** [comment]
Thank You so much :)

**Comment by rajku-dev** [comment]
Changes made in last commits -  [preview](https://deploy-preview-9192--care-ohc.netlify.app/facility/51162830-1f88-445b-8120-32ffa001a44c/assets/9ea510bb-baf6-41e9-a2df-e0115292b014/update)
- Form Validation using Zod, old validator functions (eg. EmailValidator etc) are used to refine validation
- Add dirty state to prevent form submission when there is no change
- Added new shadcn components eg form, Switch, Calendar, Select, Switch, TextArea.
-  Refactor shadcn calendar to remove error

but there are changes required in
- [x] Refactor LocationSelect to restore original functionality for all pages that uses it
- [x] Fix the layout of Calendar 
- [x] Improve CSS and styling

**Comment by rajku-dev** [comment]
@rithviknishad are these changes workable?
Asset Form now completely uses shadcn components with a dirty state.
If you have any suggestions please let me know so that i can work on it further.
Thank You.

File: src/components/Facility/AssetCreate.tsx
Comment (Jacobjeevan) [inline src/components/Facility/AssetCreate.tsx]: We are slowly migrating to using tanstack's `useQuery` and `useMutate`. Let's update for these files.

File: src/components/Facility/AssetCreate.tsx
Comment (Jacobjeevan) [inline src/components/Facility/AssetCreate.tsx]: `useQuery` here.

File: src/components/Facility/AssetCreate.tsx
Comment (Jacobjeevan) [inline src/components/Facility/AssetCreate.tsx]: Remove the console.logs 👍

File: src/components/Facility/AssetCreate.tsx
Comment (Jacobjeevan) [inline src/components/Facility/AssetCreate.tsx]: There's issues with both rendering and behavior for the calendar. If you aren't able to solve it, stick with existing DateFormField for now; we will need to migrate or create a similar component with Shadcn's styles eventually.

**Review by Jacobjeevan — CHANGES_REQUESTED** [review]
Overall, good job nonetheless, esp on sticking with the issue despite all the changes 👍

**Comment by rajku-dev** [comment]
> Overall, good job nonetheless, esp on sticking with the issue despite all the changes 👍

Thanks🙌
I hope the original issue will be reopened and re-assigned to me.

File: src/components/Facility/AssetCreate.tsx
Comment (rajku-dev) [reply src/components/Facility/AssetCreate.tsx]: I was aware of that, just needed a green signal if I'm on a right track 
made the changes as requested.

File: src/components/Facility/AssetCreate.tsx
Comment (rajku-dev) [reply src/components/Facility/AssetCreate.tsx]: `useTanStackQueryInstead` is also a wrapper around `useQuery` hook right?
I've replaced those with `useQuery` for GET and `useMutation` for POST/UPDATE requests.

File: src/components/Facility/AssetCreate.tsx
Comment (rajku-dev) [reply src/components/Facility/AssetCreate.tsx]: removed console logs as requested

File: src/components/Facility/AssetCreate.tsx
Comment (rajku-dev) [reply src/components/Facility/AssetCreate.tsx]: will be working on it..

Comment (rajku-dev) [reply src/components/Facility/AssetCreate.tsx]: > There's issues with both rendering and behavior for the calendar. If you aren't able to solve it, stick with existing DateFormField for now; we will need to migrate or create a similar component with Shadcn's styles eventually.

@jacobjeevan tried to align it properly, but i think its a problem in the template itself,  hope they fix it soon🤞
I've switched to good old `DateFormField` for Expiry and Last Service Date fields wrapped in ShadCN FormControl, please review the changes.
Thank You

**Comment by rajku-dev** [comment]
All fields look pretty similar to the original one now

image

**Comment by rithviknishad** [comment]
"assets" is no longer present in new architecture as of now. we can rethink once it's brought back.

**Pull request closed without merging**
